File: VisaCheckout/src/main/java/com/braintreepayments/api/VisaCheckoutAccount.java
package com.braintreepayments.api;

import com.visa.checkout.VisaPaymentSummary;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Use to construct a Visa Checkout tokenization request.
 */
class VisaCheckoutAccount extends PaymentMethod {
    private static final String CALL_ID = "callId";
    private static final String ENCRYPTED_KEY = "encryptedKey";
    private static final String ENCRYPTED_PAYMENT_DATA = "encryptedPaymentData";
    private static final String VISA_CHECKOUT_KEY = "visaCheckoutCard";

    private String callId;
    private String encryptedKey;
    private String encryptedPaymentData;

    /**
     * @param visaPaymentSummary returned from Visa Checkout after a successful payment.
     */
    VisaCheckoutAccount(VisaPaymentSummary visaPaymentSummary) {
        if (visaPaymentSummary == null) {
            return;
        }

        callId = visaPaymentSummary.getCallId();
        encryptedKey = visaPaymentSummary.getEncKey();
        encryptedPaymentData = visaPaymentSummary.getEncPaymentData();
    }

    @Override
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    public JSONObject buildJSON() throws JSONException {
        JSONObject json = super.buildJSON();

        JSONObject paymentMethodNonceJson = new JSONObject();
        paymentMethodNonceJson.put(CALL_ID, callId);
        paymentMethodNonceJson.put(ENCRYPTED_KEY, encryptedKey);
        paymentMethodNonceJson.put(ENCRYPTED_PAYMENT_DATA, encryptedPaymentData);

        json.put(VISA_CHECKOUT_KEY, paymentMethodNonceJson);
        return json;
    }

    @Override
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    public String getApiPath() {
        return "visa_checkout_cards";
    }
}

File: VisaCheckout/src/main/java/com/braintreepayments/api/visacheckout/VisaCheckoutAccount.kt
package com.braintreepayments.api.visacheckout

import androidx.annotation.RestrictTo
import com.braintreepayments.api.core.IntegrationType
import com.braintreepayments.api.core.MetadataBuilder
import com.braintreepayments.api.core.PaymentMethod
import com.braintreepayments.api.core.PaymentMethod.Companion.DEFAULT_SOURCE
import com.visa.checkout.VisaPaymentSummary
import org.json.JSONException
import org.json.JSONObject

/**
 * Use to construct a Visa Checkout tokenization request.
 */
internal class VisaCheckoutAccount(
    private val visaPaymentSummary: VisaPaymentSummary,
    override var sessionId: String? = null,
    override var source: String? = DEFAULT_SOURCE,
    override var integration: IntegrationType? = IntegrationType.CUSTOM
) : PaymentMethod {

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    private fun buildMetadataJSON(): JSONObject {
        return MetadataBuilder()
            .sessionId(sessionId)
            .source(source)
            .integration(integration)
            .build()
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    @Throws(JSONException::class)
    override fun buildJSON(): JSONObject {
        val json = JSONObject()
            .put(MetadataBuilder.META_KEY, buildMetadataJSON())
        val paymentMethodNonceJson = JSONObject().apply {
            put(CALL_ID, visaPaymentSummary.callId)
            put(ENCRYPTED_KEY, visaPaymentSummary.encKey)
            put(ENCRYPTED_PAYMENT_DATA, visaPaymentSummary.encPaymentData)
        }
        json.put(VISA_CHECKOUT_KEY, paymentMethodNonceJson)
        return json
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    override val apiPath = "visa_checkout_cards"

    companion object {
        private const val CALL_ID = "callId"
        private const val ENCRYPTED_KEY = "encryptedKey"
        private const val ENCRYPTED_PAYMENT_DATA = "encryptedPaymentData"
        private const val VISA_CHECKOUT_KEY = "visaCheckoutCard"
    }
}
